File: agent/src/shared/observability/metrics.ts
const MAX_LATENCY_SAMPLES = 5000;
const QPS_WINDOW_SECONDS = 60;

type MetricBucket = {
  count: number;
  errorCount: number;
  totalLatencyMs: number;
  latencySamples: number[];
};

type MetricsState = {
  startedAtMs: number;
  httpSummary: MetricBucket;
  routes: Map<string, MetricBucket>;
  downstream: Map<string, MetricBucket>;
  requestTimestamps: number[];
};

export type BucketSnapshot = {
  count: number;
  error_count: number;
  error_rate: number;
  avg_latency_ms: number;
  p95_latency_ms: number;
  p99_latency_ms: number;
};

export type MetricsSnapshot = {
  timestamp: string;
  uptime_sec: number;
  summary: {
    total_requests: number;
    total_errors: number;
    error_rate: number;
    qps_avg: number;
    qps_1m: number;
    avg_latency_ms: number;
    p95_latency_ms: number;
    p99_latency_ms: number;
  };
  routes: Record<string, BucketSnapshot>;
  downstream: Record<string, BucketSnapshot>;
};

function createBucket(): MetricBucket {
  return {
    count: 0,
    errorCount: 0,
    totalLatencyMs: 0,
    latencySamples: [],
  };
}

const state: MetricsState = {
  startedAtMs: Date.now(),
  httpSummary: createBucket(),
  routes: new Map<string, MetricBucket>(),
  downstream: new Map<string, MetricBucket>(),
  requestTimestamps: [],
};

function isErrorStatus(statusCode: number): boolean {
  return statusCode >= 400 || statusCode <= 0;
}

function addSample(bucket: MetricBucket, latencyMs: number, isError: boolean) {
  bucket.count += 1;
  if (isError) {
    bucket.errorCount += 1;
  }

  const safeLatency = Number.isFinite(latencyMs) && latencyMs >= 0 ? latencyMs : 0;
  bucket.totalLatencyMs += safeLatency;
  bucket.latencySamples.push(safeLatency);

  if (bucket.latencySamples.length > MAX_LATENCY_SAMPLES) {
    bucket.latencySamples.splice(0, bucket.latencySamples.length - MAX_LATENCY_SAMPLES);
  }
}

function percentile(samples: number[], target: number): number {
  if (samples.length === 0) {
    return 0;
  }

  const copy = [...samples].sort((a, b) => a - b);
  const rank = Math.ceil((target / 100) * copy.length) - 1;
  const index = Math.max(0, Math.min(copy.length - 1, rank));
  return copy[index] ?? 0;
}

function round(value: number, digits: number): number {
  const factor = 10 ** digits;
  return Math.round(value * factor) / factor;
}

function formatBucket(bucket: MetricBucket): BucketSnapshot {
  if (bucket.count <= 0) {
    return {
      count: 0,
      error_count: 0,
      error_rate: 0,
      avg_latency_ms: 0,
      p95_latency_ms: 0,
      p99_latency_ms: 0,
    };
  }

  const avg = bucket.totalLatencyMs / bucket.count;
  return {
    count: bucket.count,
    error_count: bucket.errorCount,
    error_rate: round(bucket.errorCount / bucket.count, 4),
    avg_latency_ms: round(avg, 2),
    p95_latency_ms: round(percentile(bucket.latencySamples, 95), 2),
    p99_latency_ms: round(percentile(bucket.latencySamples, 99), 2),
  };
}

function routeKey(method: string, route: string): string {
  return `${method.toUpperCase()} ${route}`;
}

function downstreamKey(service: string, operation: string): string {
  return `${service}:${operation}`;
}

function purgeQPSWindow(nowSec: number) {
  const cutoff = nowSec - QPS_WINDOW_SECONDS + 1;
  let idx = 0;

  while (idx < state.requestTimestamps.length && (state.requestTimestamps[idx] ?? 0) < cutoff) {
    idx += 1;
  }

  if (idx > 0) {
    state.requestTimestamps = state.requestTimestamps.slice(idx);
  }
}

export function recordHttpRequest(method: string, route: string, statusCode: number, latencyMs: number): void {
  const normalizedRoute = route || 'UNKNOWN';
  const nowSec = Math.floor(Date.now() / 1000);
  const isError = isErrorStatus(statusCode);

  addSample(state.httpSummary, latencyMs, isError);

  const key = routeKey(method, normalizedRoute);
  let bucket = state.routes.get(key);
  if (!bucket) {
    bucket = createBucket();
    state.routes.set(key, bucket);
  }
  addSample(bucket, latencyMs, isError);

  state.requestTimestamps.push(nowSec);
  purgeQPSWindow(nowSec);
}

export function recordDownstream(service: string, operation: string, statusCode: number, latencyMs: number): void {
  const normalizedService = service || 'unknown';
  const normalizedOperation = operation || 'unknown';
  const key = downstreamKey(normalizedService, normalizedOperation);

  let bucket = state.downstream.get(key);
  if (!bucket) {
    bucket = createBucket();
    state.downstream.set(key, bucket);
  }

  addSample(bucket, latencyMs, isErrorStatus(statusCode));
}

export function snapshotMetrics(): MetricsSnapshot {
  const nowMs = Date.now();
  const nowSec = Math.floor(nowMs / 1000);
  purgeQPSWindow(nowSec);

  const uptimeSec = Math.max(1, Math.floor((nowMs - state.startedAtMs) / 1000));
  const summaryBucket = formatBucket(state.httpSummary);
  const qpsAvg = state.httpSummary.count / uptimeSec;
  const qps1m = state.requestTimestamps.length / QPS_WINDOW_SECONDS;

  const routeSnapshot: Record<string, BucketSnapshot> = {};
  for (const [key, bucket] of state.routes.entries()) {
    routeSnapshot[key] = formatBucket(bucket);
  }

  const downstreamSnapshot: Record<string, BucketSnapshot> = {};
  for (const [key, bucket] of state.downstream.entries()) {
    downstreamSnapshot[key] = formatBucket(bucket);
  }

  return {
    timestamp: new Date(nowMs).toISOString(),
    uptime_sec: uptimeSec,
    summary: {
      total_requests: state.httpSummary.count,
      total_errors: state.httpSummary.errorCount,
      error_rate: summaryBucket.error_rate,
      qps_avg: round(qpsAvg, 4),
      qps_1m: round(qps1m, 4),
      avg_latency_ms: summaryBucket.avg_latency_ms,
      p95_latency_ms: summaryBucket.p95_latency_ms,
      p99_latency_ms: summaryBucket.p99_latency_ms,
    },
    routes: routeSnapshot,
    downstream: downstreamSnapshot,
  };
}

File: backend/internal/observability/metrics.go
package observability

import (
	"math"
	"sort"
	"sync"
	"time"
)

const (
	maxLatencySamples = 5000
	qpsWindowSeconds  = 60
)

type metricBucket struct {
	Count          uint64
	ErrorCount     uint64
	TotalLatencyMs float64
	LatencySamples []float64
}

type collector struct {
	mu sync.Mutex

	startedAt time.Time

	httpSummary metricBucket
	routes      map[string]*metricBucket
	downstreams map[string]*metricBucket

	requestTimestamps []int64
}

var globalCollector = &collector{
	startedAt:   time.Now(),
	routes:      make(map[string]*metricBucket),
	downstreams: make(map[string]*metricBucket),
}

type BucketSnapshot struct {
	Count        uint64  `json:"count"`
	ErrorCount   uint64  `json:"error_count"`
	ErrorRate    float64 `json:"error_rate"`
	AvgLatencyMs float64 `json:"avg_latency_ms"`
	P95LatencyMs float64 `json:"p95_latency_ms"`
	P99LatencyMs float64 `json:"p99_latency_ms"`
}

type SummarySnapshot struct {
	TotalRequests uint64  `json:"total_requests"`
	TotalErrors   uint64  `json:"total_errors"`
	ErrorRate     float64 `json:"error_rate"`
	QPSAvg        float64 `json:"qps_avg"`
	QPS1m         float64 `json:"qps_1m"`
	AvgLatencyMs  float64 `json:"avg_latency_ms"`
	P95LatencyMs  float64 `json:"p95_latency_ms"`
	P99LatencyMs  float64 `json:"p99_latency_ms"`
}

type Snapshot struct {
	Timestamp  string                    `json:"timestamp"`
	UptimeSec  int64                     `json:"uptime_sec"`
	Summary    SummarySnapshot           `json:"summary"`
	Routes     map[string]BucketSnapshot `json:"routes"`
	Downstream map[string]BucketSnapshot `json:"downstream"`
}

func addSample(bucket *metricBucket, latencyMs float64, isError bool) {
	bucket.Count++
	if isError {
		bucket.ErrorCount++
	}
	bucket.TotalLatencyMs += latencyMs
	bucket.LatencySamples = append(bucket.LatencySamples, latencyMs)
	if len(bucket.LatencySamples) > maxLatencySamples {
		overflow := len(bucket.LatencySamples) - maxLatencySamples
		bucket.LatencySamples = bucket.LatencySamples[overflow:]
	}
}

func formatBucket(bucket *metricBucket) BucketSnapshot {
	if bucket == nil || bucket.Count == 0 {
		return BucketSnapshot{}
	}

	avg := bucket.TotalLatencyMs / float64(bucket.Count)
	p95 := percentile(bucket.LatencySamples, 95)
	p99 := percentile(bucket.LatencySamples, 99)
	errorRate := float64(bucket.ErrorCount) / float64(bucket.Count)

	return BucketSnapshot{
		Count:        bucket.Count,
		ErrorCount:   bucket.ErrorCount,
		ErrorRate:    round(errorRate, 4),
		AvgLatencyMs: round(avg, 2),
		P95LatencyMs: round(p95, 2),
		P99LatencyMs: round(p99, 2),
	}
}

func percentile(samples []float64, p int) float64 {
	if len(samples) == 0 {
		return 0
	}
	copySamples := make([]float64, len(samples))
	copy(copySamples, samples)
	sort.Float64s(copySamples)

	rank := int(math.Ceil((float64(p) / 100) * float64(len(copySamples))))
	if rank < 1 {
		rank = 1
	}
	if rank > len(copySamples) {
		rank = len(copySamples)
	}

	return copySamples[rank-1]
}

func round(value float64, digits int) float64 {
	pow := math.Pow(10, float64(digits))
	return math.Round(value*pow) / pow
}

func errorStatus(statusCode int) bool {
	return statusCode >= 400
}

func routeKey(method, route string) string {
	return method + " " + route
}

func downstreamKey(service, operation string) string {
	return service + ":" + operation
}

func (c *collector) purgeQPSWindow(now int64) {
	cutoff := now - qpsWindowSeconds + 1
	idx := 0
	for idx < len(c.requestTimestamps) && c.requestTimestamps[idx] < cutoff {
		idx++
	}
	if idx > 0 {
		c.requestTimestamps = c.requestTimestamps[idx:]
	}
}

// RecordHTTPRequest 记录 HTTP 层请求指标。
func RecordHTTPRequest(method, route string, statusCode int, latency time.Duration) {
	if route == "" {
		route = "UNKNOWN"
	}

	now := time.Now().Unix()
	latencyMs := float64(latency.Milliseconds())

	globalCollector.mu.Lock()
	defer globalCollector.mu.Unlock()

	isError := errorStatus(statusCode)
	addSample(&globalCollector.httpSummary, latencyMs, isError)

	key := routeKey(method, route)
	bucket := globalCollector.routes[key]
	if bucket == nil {
		bucket = &metricBucket{}
		globalCollector.routes[key] = bucket
	}
	addSample(bucket, latencyMs, isError)

	globalCollector.requestTimestamps = append(globalCollector.requestTimestamps, now)
	globalCollector.purgeQPSWindow(now)
}

// RecordDownstream 记录下游服务调用指标（如 backend -> agent）。
func RecordDownstream(service, operation string, statusCode int, latency time.Duration) {
	if service == "" {
		service = "unknown"
	}
	if operation == "" {
		operation = "unknown"
	}

	latencyMs := float64(latency.Milliseconds())
	isError := statusCode <= 0 || errorStatus(statusCode)

	globalCollector.mu.Lock()
	defer globalCollector.mu.Unlock()

	key := downstreamKey(service, operation)
	bucket := globalCollector.downstreams[key]
	if bucket == nil {
		bucket = &metricBucket{}
		globalCollector.downstreams[key] = bucket
	}
	addSample(bucket, latencyMs, isError)
}

// SnapshotMetrics 获取当前指标快照。
func SnapshotMetrics() Snapshot {
	globalCollector.mu.Lock()
	defer globalCollector.mu.Unlock()

	now := time.Now()
	uptimeSec := int64(now.Sub(globalCollector.startedAt).Seconds())
	if uptimeSec < 1 {
		uptimeSec = 1
	}

	globalCollector.purgeQPSWindow(now.Unix())

	summary := formatBucket(&globalCollector.httpSummary)
	qpsAvg := float64(globalCollector.httpSummary.Count) / float64(uptimeSec)
	qps1m := float64(len(globalCollector.requestTimestamps)) / qpsWindowSeconds

	routes := make(map[string]BucketSnapshot, len(globalCollector.routes))
	for key, bucket := range globalCollector.routes {
		routes[key] = formatBucket(bucket)
	}

	downstream := make(map[string]BucketSnapshot, len(globalCollector.downstreams))
	for key, bucket := range globalCollector.downstreams {
		downstream[key] = formatBucket(bucket)
	}

	return Snapshot{
		Timestamp: now.Format(time.RFC3339),
		UptimeSec: uptimeSec,
		Summary: SummarySnapshot{
			TotalRequests: globalCollector.httpSummary.Count,
			TotalErrors:   globalCollector.httpSummary.ErrorCount,
			ErrorRate:     summary.ErrorRate,
			QPSAvg:        round(qpsAvg, 4),
			QPS1m:         round(qps1m, 4),
			AvgLatencyMs:  summary.AvgLatencyMs,
			P95LatencyMs:  summary.P95LatencyMs,
			P99LatencyMs:  summary.P99LatencyMs,
		},
		Routes:     routes,
		Downstream: downstream,
	}
}
